genesyscloud/routing_queue/genesyscloud_routing_queue_proxy.go

```diff
@@ -0,0 +1,163 @@
+	getAllRoutingQueuesAttr          getAllRoutingQueuesFunc
+	getRoutingQueueByIdAttr          getRoutingQueueByIdFunc
+	getRoutingQueueWrapupCodeIdsAttr getRoutingQueueWrapupCodeIdsFunc
+	RoutingQueueCache                rc.CacheInterface[platformclientv2.Queue]
+}
+
+// newRoutingQueuesProxy initializes the routing queue proxy with all the data needed to communicate with Genesys Cloud
+func newRoutingQueuesProxy(clientConfig *platformclientv2.Configuration) *RoutingQueueProxy {
+	api := platformclientv2.NewRoutingApiWithConfig(clientConfig)
+	routingQueueCache := rc.NewResourceCache[platformclientv2.Queue]()
+
+	return &RoutingQueueProxy{
+		clientConfig:                     clientConfig,
+		routingApi:                       api,
+		getAllRoutingQueuesAttr:          getAllRoutingQueuesFn,
+		getRoutingQueueByIdAttr:          getRoutingQueueByIdFn,
+		getRoutingQueueWrapupCodeIdsAttr: getRoutingQueueWrapupCodeIdsFn,
+		RoutingQueueCache:                routingQueueCache,
+	}
+}
+
+// GetRoutingQueueProxy acts as a singleton to for the internalProxy.  It also ensures
+// that we can still proxy our tests by directly setting internalProxy package variable
+func GetRoutingQueueProxy(clientConfig *platformclientv2.Configuration) *RoutingQueueProxy {
+	if internalProxy == nil {
+		internalProxy = newRoutingQueuesProxy(clientConfig)
+	}
+	return internalProxy
+}
+
+// GetAllRoutingQueues retrieves all Genesys Cloud routing queues
+func (p *RoutingQueueProxy) GetAllRoutingQueues(ctx context.Context) (*[]platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	return p.getAllRoutingQueuesAttr(ctx, p)
+}
+
+// getRoutingQueueById returns a single Genesys Cloud Routing Queue by ID
+func (p *RoutingQueueProxy) getRoutingQueueById(ctx context.Context, queueId string) (*platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	return p.getRoutingQueueByIdAttr(ctx, p, queueId)
+}
+
+// getRoutingQueueWrapupCodeIds returns a list of routing queue wrapup code ids
+func (p *RoutingQueueProxy) getRoutingQueueWrapupCodeIds(ctx context.Context, queueId string) ([]string, *platformclientv2.APIResponse, error) {
+	return p.getRoutingQueueWrapupCodeIdsAttr(ctx, p, queueId)
+}
+
+// getAllRoutingQueuesFn is the implementation for retrieving all routing queues in Genesys Cloud
+func getAllRoutingQueuesFn(ctx context.Context, p *RoutingQueueProxy) (*[]platformclientv2.Queue, *platformclientv2.APIResponse, error) {
+	var allQueues []platformclientv2.Queue
+	const pageSize = 100
+
+	queues, resp, getErr := p.routingApi.GetRoutingQueues(1, pageSize, "", "", nil, nil, nil, false)
+	if getErr != nil {
+		return nil, resp, fmt.Errorf("failed to get first page of queues: %v", getErr)
+	}
+
+	// Check if the routing queue cache is populated with all the data, if it is, return that instead
+	// If the size of the cache is the same as the total number of queues, the cache is up-to-date
+	if rc.GetCacheSize(p.RoutingQueueCache) == *queues.Total {
```

Comment:
do we need cache for getAll ?

Reply:
The reason I have it like this is because this getAll will eventually be used by 3 resources, `routing_queues`, `routing_queues_conditional_group_routing` and `routing_queue_outbound_email_address`. If I cache the get all when the first resource call it, it stops the other two resources having to make the same call